### PR TITLE
Rest entity responsive module without tests.

### PR DESCRIPTION
**Pull request by PavelBulat**

### CODE REVIEW

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name: REST Entity Recursive
+description: Adds format to REST endpoints. The request with the format includes entity with whole references entities.
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: I'd update description to:
A new "json_recursive" REST format exposes all fields and referenced entities by default.

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+package: REST
+type: module
+core: 8.x
+dependencies:
+  - consumer_image_styles
+  - paragraphs:paragraphs
+  - rest
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: - drupal:rest

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+name: REST Entity Recursive
+description: Adds format to REST endpoints. The request with the format includes entity with whole references entities.
+package: REST
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: package: Web services

File: modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function supportsNormalization($data, $format = NULL) {
+    // It is very tricky to detect if a file entity is an image or not. This is
+    // typically done using a special field type to point to this entity.
+    // However we don't have access to that in here. Besides we want this to
+    // apply when requesting a listing of file entities as well, not only via
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php]: " Besides we want this to
    // apply when requesting a listing of file entities as well, not only via
    // includes"

Is it relevant to this module?

File: modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * Creates array of image styles for image.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   Entity.
+   * @param array $context
+   *   Context.
+   */
+  protected function buildVariantValues(EntityInterface $entity, array $context = []) {
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php]: If this function was copied from another module, can you please add a reference in comment?

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+type: module
+core: 8.x
+dependencies:
+  - consumer_image_styles
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: Can we drop consumer_image_styles dependency? Not everyone will need it.

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+  - paragraphs:paragraphs
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: Can we drop paragraphs dependency? Not everyone will need it.

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml (hunk 6 above)
Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: Yes we can, but we need to remove ImageItemNormalizer from this module. Actually we don`t need it in this module.

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml (hunk 7 above)
Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: Yes we can, but we need to remove ParagraphItemNormalizer from this module. Actually we don`t need it in this module.

File: modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php (hunk 5 above)
Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php]: Normalizer removed.

File: modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php (hunk 4 above)
Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/src/Normalizer/ImageItemNormalizer.php]: This file removed.

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml (hunk 3 above)
Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: Done.

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml (hunk 2 above)
Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: done.

File: modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml (hunk 1 above)
Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/rest_entity_recursive.info.yml]: Changed.

File: modules/custom/rest_entity_recursive/src/Normalizer/ContentEntityNormalizer.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    // Create the array of normalized fields.
+    $normalized = [];
+
+    // If it is root entity set current_depth and root_entity in context.
+    if (!array_key_exists('current_depth', $context)) {
+      $context['current_depth'] = 0;
+
+      // Get max_depth from request.
+      $requestMaxDepth = (int) $context['request']->query->get('max_depth');
+      $context['max_depth'] = $requestMaxDepth ? $requestMaxDepth : $this->defaultMaxDepth;
```
Comment (kalabro) [inline modules/custom/rest_entity_recursive/src/Normalizer/ContentEntityNormalizer.php]: Request with `?max_depth=0` works as max_depth=10.

Comment (PavelBulat) [reply modules/custom/rest_entity_recursive/src/Normalizer/ContentEntityNormalizer.php]: Fixed. Also fixed max_depth=-1

File: tests/api/rest_entity_recursive/RestEntityRecursiveCest.php [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  /**
+   * Checks location in redirect response.
+   *
+   * @param \ApiTester $I
+   * @group additional
+   */
+  public function testRedirectJsonRecursiveFormat(\ApiTester $I) {
+    drupal_flush_all_caches();
```
Comment (kalabro) [inline tests/api/rest_entity_recursive/RestEntityRecursiveCest.php:80]: @PavelBulat can you please add a comment explaining why we need drupal_flush_all_caches() here?

**Review by kalabro — APPROVED** [review]

**Pull request merged**